src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.cs

```diff
@@ -115,44 +160,57 @@ private CborInitialByte PeekInitialByte(CborMajorType expectedType)
             return result;
         }
 
+        private void ReadNextIndefiniteLengthBreakByte()
+        {
+            CborInitialByte result = PeekInitialByte();
+
+            if (result.InitialByte != CborInitialByte.IndefiniteLengthBreakByte)
+            {
+                throw new InvalidOperationException("Next data item is not indefinite-length break byte.");
+            }
+        }
+
         private void PushDataItem(CborMajorType type, ulong? expectedNestedItems)
         {
             if (expectedNestedItems > (ulong)_buffer.Length)
             {
                 throw new FormatException("Insufficient buffer size for declared definite length in CBOR data item.");
             }
 
-            _nestedDataItemStack ??= new Stack<(CborMajorType, ulong?)>();
-            _nestedDataItemStack.Push((type, _remainingDataItems));
+            _nestedDataItemStack ??= new Stack<(CborMajorType, bool, ulong?)>();
+            _nestedDataItemStack.Push((type, _isEvenNumberOfDataItemsWritten, _remainingDataItems));
             _remainingDataItems = expectedNestedItems;
+            _isEvenNumberOfDataItemsWritten = true;
         }
 
         private void PopDataItem(CborMajorType expectedType)
         {
-            if (_remainingDataItems == null)
-            {
-                throw new NotImplementedException("Indefinite-length data items");
-            }
-
-            if (_remainingDataItems > 0)
-            {
-                throw new InvalidOperationException("Definite-length nested CBOR data item is incomplete.");
-            }
-
             if (_nestedDataItemStack is null || _nestedDataItemStack.Count == 0)
             {
                 throw new InvalidOperationException("No active CBOR nested data item to pop");
             }
 
-            (CborMajorType actualType, ulong? remainingItems) = _nestedDataItemStack.Peek();
+            (CborMajorType actualType, bool isEvenNumberOfDataItemsWritten, ulong? remainingItems) = _nestedDataItemStack.Peek();
 
             if (expectedType != actualType)
             {
                 throw new InvalidOperationException("Unexpected major type in nested CBOR data item.");
             }
 
+            if (_remainingDataItems > 0)
```

Comment:
Why move this down? It seems generally cheaper to deal with than stack popping

Reply:
I felt that reporting a type error was more significant than potentially missing elements.